node/src/components/storage/disjoint_sequences.rs

```diff
@@ -0,0 +1,314 @@
+use std::fmt::{self, Display, Formatter};
+
+use datasize::DataSize;
+use itertools::Itertools;
+use tracing::trace;
+
+/// The outcome of an attempt to insert a value into a `Sequence`.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum InsertOutcome {
+    /// The value was greater than `Sequence::high + 1` and wasn't inserted.
+    TooHigh,
+    /// The value was inserted at the high end, and is now `Sequence::high`.
+    ExtendedHigh,
+    /// The value was a duplicate; inserted and didn't affect the high or low values.
+    AlreadyInSequence,
+    /// The value was inserted at the low end, and is now `Sequence::low`.
+    ExtendedLow,
+    /// The value was less than `Sequence::low - 1` and wasn't inserted.
+    TooLow,
+}
+
+/// Represents a continuous sequence of `u64`s.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, DataSize, Ord, PartialOrd)]
+pub(super) struct Sequence {
+    /// The upper bound (inclusive) of the sequence.
+    high: u64,
+    /// The lower bound (inclusive) of the sequence.
+    low: u64,
+}
+
+impl Sequence {
+    /// Constructs a new sequence containing only `value`.
+    fn new(value: u64) -> Self {
+        Sequence {
+            high: value,
+            low: value,
+        }
+    }
+
+    #[cfg(test)]
+    pub(crate) fn new_with_bounds(low: u64, high: u64) -> Self {
+        Sequence { high, low }
+    }
+
+    /// Tries to insert `value` into the sequence.
+    ///
+    /// Returns an outcome which indicates where the value was inserted if at all.
+    fn try_insert(&mut self, value: u64) -> InsertOutcome {
+        if value == self.high + 1 {
+            self.high = value;
+            InsertOutcome::ExtendedHigh
+        } else if value >= self.low && value <= self.high {
+            InsertOutcome::AlreadyInSequence
+        } else if value + 1 == self.low {
+            self.low = value;
+            InsertOutcome::ExtendedLow
+        } else if value > self.high {
+            InsertOutcome::TooHigh
+        } else {
+            InsertOutcome::TooLow
+        }
+    }
+}
+
+/// Represents a collection of disjoint sequences of `u64`s.
+///
+/// The collection is kept ordered from high to low, and each entry represents a discrete portion of
+/// the space from [0, u64::MAX] with a gap of at least 1 between each.
+///
+/// The collection is ordered this way to optimize insertion for the normal use case: adding
+/// monotonically increasing values representing the latest block height.
+///
+/// As values are inserted, if two separate sequences become contiguous, they are merged into a
+/// single sequence.
+///
+/// For example, if `sequences` contains `[9,9], [7,3]` and `8` is inserted, then `sequences` will
+/// be reduced to `[9,3]`.
+#[derive(Default, Debug, DataSize)]
+pub(super) struct DisjointSequences {
+    sequences: Vec<Sequence>,
```

Comment:
Wouldn't a bloom filter be equivalent in terms of functionality but be faster and occupy far less space? We might get a _false positive_ occasionally which could lead to user requesting a block we don't have but this should be rare and I think we can provide a human-readable response in those cases.

Reply:
I think filtering is not the main point here, though, is it? The point here isn't to remember which blocks have bodies, but which blocks with bodies form continuous sequences, and a Bloom filter wouldn't be able to do that.

Reply:
Yes, I agree with Bart, and also I'm not sure in the common case (where I expect we have very few actual disjoint sequences) that a bloom filter would be faster or occupy less space.